tsdb/querier.go

```diff
@@ -609,10 +611,16 @@ func (p *populateWithDelChunkSeriesIterator) Next() bool {
 
 	p.curr = p.currChkMeta
 	if p.currDelIter == nil {
-		return true
+		// Select ensures the return bytes data are accessible to user, so we need to copy the bytes that are mmapped at this point.
```

Comment:
This isn't how we handle this elsewhere, instead we make sure we don't deallocate until things are closed.

If a caller needs a copy it should be done by the caller, not in general code like this.

Reply:
I thought about it and I decided that it's better to do it inside the API. The problem is that in many cases this data is copied (e.g from memSeries) already, so copying on caller side, is:

* In many cases duplicated alloc as memory is safe
* You dont' know what exactly to copy. Whole response? E.g how do i Know if strings behind labels are safe to use or should I copy them?
* Easy to miss, hard to test against.

What do you think? (: @brian-brazil

Reply:
I think we should be consistent, and leave it entirely up to the caller like we do in all other APIs.

Which are the callers having the issue here? Why can't they call Close at the right time?

Reply:
Yea, we cannot because we reuse StoreAPI internally and we don't want to leak Close for this - we don't control the exact moment for serialization ):

Reply:
or worse - copying bytes to TCP packet ;p

Reply:
so at end we would need to copy bytes all perform deep copy of whole response going with this idea "caller is  responsible" which is extremely hard to  do

Plus, Why you say it's consistent? Select for samples iterator does not requires you to copy anything on caller side :thinking:

Reply:
> Yea, we cannot because we reuse StoreAPI internally and we don't want to leak Close for this - we don't control the exact moment for serialization ):

Where is this in Prometheus? We already have the plumbing for this in the API.

> Select for samples iterator does not requires you to copy anything on caller side

Yes it does, the strings for the label values can disappear after the Close.

Reply:
> Where is this in Prometheus? We already have the plumbing for this in the API.

It's on Thanos: https://github.com/thanos-io/thanos/pull/3046/files#diff-be92e59b0e37723a0a0be5cebd88d264R456

> Yes it does, the strings for the label values can disappear after the Close.

Technically they are mmaped indeed... but somehow it's hard to repro segfault on those, I would be not surprised if we copy those strings accidentally somewhere. That's the point - it's so easy to miss and extra allocate.

OK, in this case I guess I need to remove copying here, and check how I can maintain open querier across internal StoreAPI call. Let' me adjust PR:

* new requirement / note in comments about this behaviour
* adjust tests for expecting segfault only when closed is invoked on those.

Reply:
Adjusted all.